sound/soc/sof/core.c

```diff
@@ -224,12 +224,12 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 	if (ret < 0) {
 		dev_err(sdev->dev,
 			"error: failed to register DSP DAI driver %d\n", ret);
-		goto fw_run_err;
+		goto fw_trace_err;
```

Comment:
related to this: at line 203:
```
		sdev->dtrace_is_supported = true;
		/* init DMA trace */
		ret = snd_sof_init_trace(sdev);
		if (ret < 0) {
```
looks wrong to me too. We shouldn't set `sdev->dtrace_is_supported` unless `snd_sof_init_trace()` completed successfully. Otherwise we'll try to free what hasn't been allocated.

Reply:
@lyakh Ack, you are right. Will fix in next revision.

Reply:
@kv2019i indeed, `snd_sof_init_trace()` needs `dtrace_is_supported`, but what if the first call to `snd_dma_alloc_pages()` in it fails. `dtrace_is_supported` will stay on, so when `snd_sof_free_trace()` is called, it will call `snd_dma_free_pages()`. Is that ok?

Reply:
@lyakh I think calling snd_dma_free_pages() is safe at least for the sgbufs we use as they track the allocation state themselves. But I guess we can make this explicit as well. Added one new patch to the series addressing this.